### DISCUSSION TITLE
F' Serial Port Type

### DISCUSSION THREAD

**Discussion by teddylambert** [nasa/fprime#1099] [Q&A] [answered]
In #1036 the idea of the Serial port type in F' got brought up, and I wanted to bring that point into a new discussion to get some more clarification.

@LeStarch's answer was:
 > The "Serial" data type is a special data type that just reveals that buffer as a byte array. This means you can treat an input from any component's output port as a generic array of bytes in the receiving component.

Let's say I have a component that wants to send a vector of <x,y,z> data, all F32, to the generic hub. In order to do so, would I need to create a primitive float array of these three values, and then pass them into an `ExternalSerializeBuffer` with something like this `Fw::ExternalSerializeBuffer wrapper(rawData, rawSize)` (where `rawData` is that primitive array, and `rawSize` is the size of that array)?

And then the topology would look something like:
```
<connection name="VectorMsgToHub">
      <source component="someComponent" port="VectorMsgOut" type="Serial" num="0"/>
      <target component="genericHub" port="portIn" type="Serial" num="0"/>
</connection>
```

I also have access to a custom port type for this vector that sends the three F32 values as separate args. Is there something I could change with that port definition to accomplish the same thing as above? Is there one method that is recommended over the other?

**Comment by LeStarch** [comment]
Serial types can attach to anything.  Thus you can attach an F32Triple port as described above directly to the serial port.  No need to serialize data yourself.  This is the utility of Serial ports, they pass the already present data serialized buffer through untouched.

Topology looks like the following.  I used the port type name `F32Triple` which would be defined as a port of 3 F32s as you described above.

```
<connection name="VectorMsgToHub">
      <source component="someComponent" port="F32Triple" type="Serial" num="0"/>
      <target component="genericHub" port="portIn" type="Serial" num="0"/>
</connection>
```
As you can see, you don't need to change your port definition at all.  Just hook it right up.  The system then does the following:

1. Call to F32Triple_out(F1, F2, F2) serializes data to a buffer just like is done with `async` ports.  No action needed from user.
2. Buffer is handed in as-is to the input port handler. https://github.com/nasa/fprime/blob/37b4bfec15837679a138ce87ff82a87ac42a3729/Svc/GenericHub/GenericHubComponentImpl.cpp#L111

You can use serial ports in the opposite direction too.
```
<connection name="VectorMsgToHub">
      <source component="genericHub" port="portIn" type="Serial" num="0"/>
      <target component="someComponent" port="F32Triple" type="Serial" num="0"/>
</connection>
```
Here the buffer you get will be automatically deserialized into our 3 F32 arguments.  https://github.com/nasa/fprime/blob/37b4bfec15837679a138ce87ff82a87ac42a3729/Svc/GenericHub/GenericHubComponentImpl.cpp#L99

The GenericHub ignores the content inside the buffer, adds metadata, sends it, and the other hub removes the metadata and sends that out another connection.  Thus it is just like an asynchronous port call, but spans an address.

**Reply by teddylambert** [reply]
So just to confirm:
- In the sending component's topology, I would define a port `F32Triple` like 
```
<port name="F32Triple" data_type="<custom defined data type>" kind="output" max_number="1">
</port>
```
- In the deployment topology, I would define the connection as
```
<connection name="VectorMsgToHub">
      <source component="sendingComponent" port="F32Triple" type="Serial" num="0"/>
      <target component="genericHub" port="portIn" type="Serial" num="0"/>
</connection>
```
where here the type is `Serial`, not `<custom defined type>` as in the component topology?

**Comment by LeStarch** [comment, marked as answer]
Not quite.  I believe it is loke this:

```
<connection name="VectorMsgToHub">
      <source component="sendingComponent" port="F32Triple" type="<custom data type>" num="0"/>
      <target component="genericHub" port="portIn" type="Serial" num="0"/>
</connection>
```

Only one side of the connection is serial.
